server/resp/src/main/java/org/infinispan/server/resp/commands/cluster/SLOTS.java

```diff
@@ -89,32 +88,36 @@ private static CompletionStage<CharSequence> getSlotsInformation(Resp3Handler ha
       return requestNodesNetworkInformation(hash.getMembers(), handler)
             .thenApply(information -> {
                StringBuilder builder = new StringBuilder();
-
-               Map<List<Address>, IntSet> segmentOwners = new HashMap<>();
-               for (int i = 0; i < hash.getNumSegments(); i++) {
-                  segmentOwners.computeIfAbsent(hash.locateOwnersForSegment(i), ignore -> IntSets.mutableEmptySet())
-                        .add(i);
-               }
-
                int size = 0;
-               for (Map.Entry<List<Address>, IntSet> entry : segmentOwners.entrySet()) {
-                  List<Address> owners = entry.getKey();
-                  IntSet segments = entry.getValue();
-
-                  for (int i = segments.nextSetBit(0); i >= 0; i = segments.nextSetBit(i + 1)) {
-                     int runStart = i;
-                     while (segments.contains(i + 1)) {
-                        i++;
-                     }
 
-                     size++;
-                     builder.append('*').append(2 + owners.size()).append(CRLF_STRING);
-                     builder.append(':').append(runStart).append(CRLF_STRING);
-                     builder.append(':').append(i).append(CRLF_STRING);
-                     for (Address owner: owners) {
-                        String info = information.get(owner);
-                        builder.append(info);
+               SegmentSlotRelation ssr = handler.respServer().segmentSlotRelation();
+
+               // These two will always be initialized in first loop below
+               int previousOwnedSegment = -1;
+               List<Address> ownersForSegment = null;
+               IntSet segments = IntSets.mutableEmptySet();
+
+               int totalSegmentCount = hash.getNumSegments();
+               for (int i = 0; i < totalSegmentCount; ++i) {
+                  List<Address> currentOwners = hash.locateOwnersForSegment(i);
+                  segments.addAll(ssr.segmentToSlots(i));
+
+                  if (!currentOwners.equals(ownersForSegment) || i == (totalSegmentCount - 1)) {
+                     if (ownersForSegment != null) {
+                        builder.append('*').append(2 + ownersForSegment.size()).append(CRLF_STRING);
+                        int start = segments.nextSetBit(0);
+                        int end = segments.lastSetBit();
```

Comment:
Since the segments are contiguous, is this safe to assume?

Reply:
Not sure I understand what you mean here. This only works because it is contiguous. Sadly `segments` should really be called `slots` to prevent confusion. But this iterates over the segments until it either completes or finds the owners mismatch. Once this happens it prints out that range as a single contiguous group of slots and then clears the collections.